data/src/main/java/org/apache/iceberg/MetricsAppender.java

```diff
@@ -0,0 +1,509 @@
+        Schema schema, MetricsCollector metricsCollector) {
+      return metricsCollector;
+    }
+
+    @Override
+    public MetricsCollector struct(Types.StructType struct, List<MetricsCollector> fieldResults) {
+      return new MetricsCollectorRecord(fieldResults);
+    }
+
+    @Override
+    public MetricsCollector field(Types.NestedField field, MetricsCollector fieldResult) {
+      return fieldResult;
+    }
+
+    @Override
+    public MetricsCollector list(Types.ListType list, MetricsCollector elementResult) {
+      return new MetricsCollectorList(elementResult);
+    }
+
+    @Override
+    public MetricsCollector map(Types.MapType map, MetricsCollector keyResult, MetricsCollector valueResult) {
+      return new MetricsCollectorMap(keyResult, valueResult);
+    }
+
+    @Override
+    public MetricsCollector primitive(Type.PrimitiveType primitive) {
+      // TODO check if comparators require previous conversion (i.e. LocalDateTime to long)?
+      switch (primitive.typeId()) {
+        case BOOLEAN:
+          return new MetricsCollectorBoolean(currentField.fieldId(), Comparators.forType(primitive));
+        case INTEGER:
+          return new MetricsCollectorInt(currentField.fieldId(), Comparators.forType(primitive));
+        case LONG:
+          return new MetricsCollectorLong(currentField.fieldId(), Comparators.forType(primitive));
+        case FLOAT:
+          return new MetricsCollectorFloat(currentField.fieldId(), Comparators.forType(primitive));
+        case DOUBLE:
+          return new MetricsCollectorDouble(currentField.fieldId(), Comparators.forType(primitive));
+        case DATE:
+          return new MetricsCollectorDate(currentField.fieldId(), Comparators.forType(primitive));
+        case TIME:
+          return new MetricsCollectorTime(currentField.fieldId(), Comparators.forType(primitive));
+        case TIMESTAMP:
+          Types.TimestampType timestamp = (Types.TimestampType) primitive;
+          return new MetricsCollectorDateTime(currentField.fieldId(),
+              Comparators.forType(primitive), timestamp.shouldAdjustToUTC());
+        case STRING:
+          return new MetricsCollectorString(currentField.fieldId(), Comparators.forType(primitive));
+        case UUID:
+          return new MetricsCollectorUUID(currentField.fieldId(), Comparators.forType(primitive));
+        case FIXED:
+          Types.FixedType fixed = (Types.FixedType) primitive;
+          return new MetricsCollectorFixed(currentField.fieldId(), Comparators.forType(primitive), fixed.length());
+        case BINARY:
+          return new MetricsCollectorBytes(currentField.fieldId(), Comparators.forType(primitive));
+        case DECIMAL:
+          Types.DecimalType decimal = (Types.DecimalType) primitive;
+          return new MetricsCollectorDecimal(currentField.fieldId(),
+              Comparators.forType(primitive), decimal.precision(), decimal.scale());
+        default:
+          return null; /* TODO */
```

Comment:
Most of the time, we throw UnsupportedOperationException in this case with a helpful error message that identifies the type ID that wasn't supported.